### ISSUE TITLE
[MNT]: Streamline mpltoolkits.axes_grid.Grid API

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#27963]
### Summary

The current interface is neither very intuitive nor practical nor in line with the traditional gridspec logic.

### Proposed fix

Half a note to self, because this is not completely thought through:

- The parameter `ngrids` seems not to be working (usage raises)
- The attributes `ngrids` is not helpful. It's misnomed (naxes would be better), and it's equivalent to `len(grid)` or `len(grid.axes_all)`.
- Not sure what we need `grid.axes_llc` for. That seems quite specific.
- The 2D grid arrangement `axes_row`, `axes_column` are lists of lists. In contrast `subplots` returns an array. Thus, you need double indexing `grid.axes_row[row][col}` and cannot use the multi-index `axs[row, col]`.
- Having the `direction` parameter and `grid.axes_row`, `grid.axes_column` is basically two redundant approaches towards ordering.

Possible improvements:
- deprecate `ngrids` attribute
- add an attribute `grid.axes` or maybe even  `grid.axs` (in analogy to `fig, axs = plt.subplots(2, 3)`), which is a ndarray of the Axes. Hence one could do `grid.axs[row, column]`.
- maybe (??) soft-deprecate axes_all, axes_row, axes_col, direction?
-  in `Grid.__init__` make everything after `nrows_ncols` kw-only
- would be nice to have a default `rect`, but that requires reorganizing parameter order because that's before `nrows_ncols`.
  Might be possible migrate
  ```
  def __init__(self, fig, rect, nrows_ncols, *, ...)
  ```
  to 
  ```
  def __init__(self, nrows_ncols, *, rect, ...)
  ```
  with reasonable deprecation warnings and catching inappropriate use. - But would have to think this through

**Comment by jklymak** [comment]
Is there a reason to cultivate this interface at all?  Isn't it superseded by features of modern matplotlib?

**Comment by timhoffm** [comment]
Is it? If so, we should at least add a discouraged note with replacement recommendations. But I believe we have no replacement at least for the ImagGrid subclass (which is from where I came here).

**Comment by jklymak** [comment]
It's probably not exactly the same, but `fig, axs = plt.subplots(2, 2, layout='compressed')` will give you something very similar and more idiomatic.

**Comment by timhoffm** [comment]
Yes, this helps with the layout. But AFAIK, the shared colorbar handling is not reasonably developed.

**Comment by timhoffm** [comment]
On a general note, I’d be happy to get rid of axes_grid1, because having two different interfaces is  awful user experience and additional maintenance cost. However, I feel this is a longer way and it may still be worth to improve on the existing structures.

**Comment by jklymak** [comment]
> Yes, this helps with the layout. But AFAIK, the shared colorbar handling is not reasonably developed.

```
fig, axs = plt.subplots(2, 2, layout='compressed')

for i in range(4):
    pc = axs.flat[i].imshow(np.random.randn(30, 30))
fig.colorbar(pc, ax=axs)

plt.show()
```
![Test](https://github.com/matplotlib/matplotlib/assets/1562854/83b36113-3bcf-4f48-a113-38cb5a1826a6)

**Comment by jklymak** [comment]
> I’d be happy to get rid of axes_grid1, because having two different interfaces is awful user experience and additional maintenance cost. However, I feel this is a longer way and it may still be worth to improve on the existing structures.

I agree with this, but I'm not sure it's worth trying to improve what is there versus improving the main library to do the same things.

**Comment by timhoffm** [comment]
The major issue with the above plots is that the colorbar is only associated with one image. So if you don't set the same explicit limits on each imshow and rely on auto-scaling, the same color may represent different values in different images. 

I thought `ImageGrid` was doing something cleverer here. But it turns out, I've been doing this in my personal wrapper function on `ImageGrid` to make it usable :upside_down_face:.

**Comment by jklymak** [comment]
Yeah, I'm not aware that `ImageGrid` somehow links the norms.  For sure that would be useful in some contexts, but is probably better handled by the user or a wrapper (I do this all the time as well, though usually for pcolormesh)

**Comment by anntzer** [comment]
> The parameter ngrids seems not to be working (usage raises)

From a quick look I appear to have broken that in #15639; a patch may be
```patch
diff --git i/lib/mpl_toolkits/axes_grid1/axes_grid.py w/lib/mpl_toolkits/axes_grid1/axes_grid.py
index 720d985414..5ea1714871 100644
--- i/lib/mpl_toolkits/axes_grid1/axes_grid.py
+++ w/lib/mpl_toolkits/axes_grid1/axes_grid.py
@@ -144,9 +144,9 @@ class Grid:
             axes_array[row, col] = axes_class(
                 fig, rect, sharex=sharex, sharey=sharey)
         self.axes_all = axes_array.ravel(
-            order="C" if self._direction == "row" else "F").tolist()
-        self.axes_column = axes_array.T.tolist()
-        self.axes_row = axes_array.tolist()
+            order="C" if self._direction == "row" else "F").tolist()[:ngrids]
+        self.axes_column = [[ax for ax in col if ax] for col in axes_array.T]
+        self.axes_row = [[ax for ax in row if ax] for row in axes_array]
         self.axes_llc = self.axes_column[0][-1]
 
         self._init_locators()
@@ -256,7 +256,10 @@ class Grid:
             return
         for i in range(self._nrows):
             for j in range(self._ncols):
-                ax = self.axes_row[i][j]
+                try:
+                    ax = self.axes_row[i][j]
+                except IndexError:
+                    continue
                 if isinstance(ax.axis, MethodType):
                     bottom_axis = SimpleAxisArtist(ax.xaxis, 1, ax.spines["bottom"])
                     left_axis = SimpleAxisArtist(ax.yaxis, 1, ax.spines["left"])
 ```

**Comment by anntzer** [comment]
Looking at this again, if you really want to redo the API I would perhaps suggest completely dropping the Grid and ImageGrid classes in favor of grid() and image_grid() functions that simply return 2D arrays of axes (similarly to subplots), i.e. the old axes_row attribute (but as a 2D ndarray).
That means you can't update certain features a posteriori anymore (set_aspect, set_label_mode, set_axes_locator) but I wouldn't worry too much about that; that's likely even much less used than the Grid/ImageGrid classes themselves.

**Comment by timhoffm** [comment]
> simply return 2D arrays of axes (similarly to subplots)

How is that different from `plt.subplots(2, 2, layout='compressed')` (c.f. https://github.com/matplotlib/matplotlib/issues/27963#issuecomment-2015664326)?
